### ISSUE TITLE
Preserving functions as values.

### ISSUE THREAD

**Issue by jgrund** [fable-compiler/Fable#965]
Consider:

```
open Fable.Core

let add a b = a + b

add
```

This emits:

```
export function add(a, b) {
  return a + b | 0;
}

(function (a, b) {
  return add(a, b);
});
````

Since functions are first class values in JS, this removes a bunch of possibilities.

Is it possible to detect functions being used as values and not wrap them?

**Comment by alfonsogarciacaro** [comment]
This is something that is done by the F# compiler (not Fable) and is very complicated to modify without affecting the semantics and opening the door to new bugs. The F# compiler does this because module and class methods are always implemented with multiple arguments but when used as a value, F# wraps them to match the semantics of F# function arguments (curried or tupled). In your example it looks the same because of the uncurrying optimization in Fable but in F# AST it would be represented like:

```js
function (a) {
    return function (b) {
        return add(a, b);
    }
}
```

The wrapping is necessary also to capture this when using types. Consider:

```js
class MyClass {
    constructor(x) {
       this.x = x;
    }
    add(y) {
       return this.x + y;
    }
}

const c = MyClass(5);
anotherObject.callback(c.add)
```

This will fail in JS because when `c.add` is called `this` will have a different meaning (because JS!) so wrapping (with an arrow function) is necessary to capture `this`.

In conclusion, from the top of my head the only place where we could apply the optimization would be for module methods used as values without tuple arguments. Of course, that will be nice to have, but right now I don't know how easy would be to do it and how much it will complicate Fable code (as a comparison, the uncurrying optimization has been a real beast with edge cases that still pop up from time to time), so I wouldn't consider it a priority at the moment.

**Comment by jgrund** [comment]
Understood. I'm not looking to open up a can of worms right now.
